src/functions.js

```diff
@@ -25,144 +41,113 @@ function setFunctionType(block, ...types) {
   block.outputType = outputType;
 }
 
-Blockly.Blocks['even'] = {
-  init: function () {
-    this.appendValueInput("NAME")
-      .appendField("even");
-    this.setInputsInline(true);
-    this.setOutput(true);
-    this.setColour(230);
-    this.setTooltip("");
-    this.setHelpUrl("");
-    this.setOnChange(onChangeFunction.bind(this))
-
-    setFunctionType(this, "Number", "Boolean")
-  }
-}
-
-Blockly.Blocks['not'] = {
-  init: function () {
-    this.appendValueInput("NAME")
-      .appendField("not");
-    this.setInputsInline(true);
-    this.setOutput(true);
-    this.setColour(230);
-    this.setTooltip("");
-    this.setHelpUrl("");
-    this.setOnChange(onChangeFunction.bind(this))
-
-    setFunctionType(this, "Boolean", "Boolean")
-  }
-}
-
-Blockly.Blocks['length'] = {
-  init: function () {
-    this.appendValueInput("NAME")
-      .appendField("length");
-    this.setInputsInline(true);
-    this.setOutput(true);
-    this.setColour(230);
-    this.setTooltip("");
-    this.setHelpUrl("");
-    this.setOnChange(onChangeFunction.bind(this))
-
-    setFunctionType(this, "String", "Number")
+function buildFuctionBlockWith(name, functionType, cb) {
+  Blockly.Blocks[name] = {
+    init: function () {
+      cb(this)
+      this.setInputsInline(true)
+      this.setOutput(true)
+      this.setTooltip("")
+      this.setHelpUrl("")
+      this.setOnChange(onChangeFunction.bind(this))
+      setFunctionType(this, ...functionType)
+    }
   }
-}
-
-Blockly.Blocks['charAt'] = {
-  init: function () {
-    this.appendValueInput("NAME")
-      .appendField("charAt");
-
-    this.appendValueInput("NAME")
 
-    this.setInputsInline(true);
-    this.setOutput(true);
-    this.setColour(230);
-    this.setTooltip("");
-    this.setHelpUrl("");
-    this.setOnChange(onChangeFunction.bind(this))
-
-    setFunctionType(this, "Number", "String", "String")
-  }
 }
 
-Blockly.Blocks['compare'] = {
+const buildFuctionBlock = (name, functionType, fields = []) =>
+  buildFuctionBlockWith(name, functionType, block => {
+    block.appendValueInput(`ARG0`).appendField(fields[0] === undefined ? name : fields[0])
+    for (let index = 1; index < functionType.length - 1; index++) {
+      const inputName = fields[index] || ""
+      block.appendValueInput(`ARG${index}`).appendField(inputName)
+    }
+  })
+
+const buildInfixFuctionBlock = ([name, field], functionType) =>
+  buildFuctionBlockWith(name, functionType, block => {
+    block.appendValueInput("LEFT")
+    block.appendValueInput("RIGHT")
+      .appendField(field)
+  })
+
+const newListType = (elementType) => new ListType(createType(elementType))
+
+buildFuctionBlock("even", ["Number", "Boolean"])
```

Comment:
I had an alternative way of implementing these, maybe there's some things that can be taken from there (like what I commented above) :

```javascript
function defineHaskellFunction(name, types, initExtension = () => {}) {
  const isPrefix = (name) => name[0].match(/[a-z]/i)
  Blockly.Blocks[name] = {
    init: function() {
      types.slice(0, -1).forEach((_type, i) => this.appendValueInput("param" + i));
      if(isPrefix(name)) {
        this.inputList[0].appendField(name)
      } else {
        this.inputList[1].appendField(name)
      }
      this.setInputsInline(true);
      this.setOutput(true);
      this.setOnChange(onChangeFunction.bind(this));

      setFunctionType(this, ...types)
    }
  }

  decorateInit(Blockly.Blocks[name], initExtension);
}

function decorateInit(block, initExtension) {
  const oldInit = block.init
  function newInit() {
    oldInit.bind(this)();
    initExtension.bind(this)();
  }
  block.init = newInit
}

defineHaskellFunction('charAt', ["Number", "String", "String"]);

defineHaskellFunction('>', ["a", "a", "Boolean"]);

Blockly.Blocks['compare'] = Blockly.Blocks['>'];
```

Reply:
Esto (y el comentario anterior) no llegué a mergearlo (habría que ver algunos detalles). 
Igual no sé bien cómo lo estás usando / necesitás para lo que estás haciendo. Las definiciones actuales ya las pasé, después fijate qué te faltaría y cualca cambialo sin problemas.